lib/tree.rb

```diff
@@ -1,66 +1,135 @@
 class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
-
-   def initialize(key, val)
+  
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
   def initialize
     @root = nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(log (n))
+  # Space Complexity: O(log (n))
   def add(key, value)
-    raise NotImplementedError
+    #recursive
+    @root = add_helper(@root, key, value)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil?
+    
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    
+    return current_node
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def find(key)
```

Comment:
If the tree is balanced this is O(log n) for both time & space and O(n) if it's not.